db/create.go

```diff
@@ -0,0 +1,156 @@
+// SQLite Caching and Cross-Exploit Database
+//
+// The db package contains the logic to handle a user provided SQLite DB in order
+// to store results and cache HTTP responses. This has a few useful benefits:
+//
+//  1. When scanning with hundreds of go-exploit implementations, the user significantly
+//     cuts down on network requests (therefore speeding up scanning), both from the
+//     verified results being cached (you only have to verify a target is Confluence once)
+//     and from the cached HTTP responses.
+//  2. The result is a useful asset database containing IP, port, installed software, and
+//     versions.
+//  3. The database can be reused with a go-exploit and generate no network traffic (assuming
+//     you aren't doing the exploitation stage). That is very interesting when, for example,
+//     you wrote a version scanner for CVE-2024-31982, scanned a customer host that was patched,
+//     but then CVE-2024-31983 is released the next day. You can essentially rescan the cached
+//     version of their system with your new CVE scanner.
+//
+// Mostly this package should be totally transparent to users of the framework. The only direct
+// interface, currently, should be calls to HTTPGetCache.
+package db
+
+import (
+	"database/sql"
+	"time"
+
+	"github.com/vulncheck-oss/go-exploit/output"
+	// pure go sqlite3 driver.
+	_ "modernc.org/sqlite"
+)
+
+// GlobalSQLHandle is a handle to the SQLite DB for handling cross-exploit data sharing.
+var GlobalSQLHandle *sql.DB
+
+// GlobalHTTPRespCacheLimit is the maximum size of an HTTP body that we will attempt to cache.
+var GlobalHTTPRespCacheLimit int
+
+const (
+	schemaVersion = "1.0.0"
+
+	metadataTable     = `CREATE TABLE IF NOT EXISTS metadata (created INTEGER NOT NULL,schema_version TEXT NOT NULL);`
+	initMetadataTable = `INSERT INTO metadata (created, schema_version) VALUES (?, ?)`
+	checkMetadata     = `SELECT name FROM sqlite_master WHERE type ='table' = ? AND name='metadata`
+	getSchemaVersion  = `SELECT schema_version FROM metadata;`
+
+	verifiedTable = `CREATE TABLE IF NOT EXISTS verified(` +
+		`id INTEGER PRIMARY KEY AUTOINCREMENT,` +
+		`created INTEGER NOT NULL,` +
+		`software_name TEXT NOT NULL,` +
+		`installed BOOLEAN NOT NULL,` +
+		`version TEXT NOT NULL,` +
+		`rhost TEXT NOT NULL,` +
```

Comment:
Do we want to add `lhost`/`lport` caching too? might be useful for situations where you need to cross-ref anycast or regional data?

Reply:
Can you expand on that? I can't think of a situation where lhost would be important, but it does sound like you have something in mind that I don't quite grasp.

Reply:
My theory was that if you were running scanners from multiple distributed nodes for workload sharing that you might aggregate the databases and having access to which host ran what could potentially be useful.

For instance, CDNs and folks with anycast networks might have different data for the same IP based on BGP route "closeness" and there might be the want to test for that. In that situation you might want access to `Lhost` to tell which node ran the scan to x-ref with region or anycast target.

Reply:
:thinking: 

That's a very interesting edge case. I think storing lhost makes sense in this context.

Reply:
Ah. One interesting tidbit is that I don't actually have access to lhost in http. I could plumb it through but I think I'd rather punt that into a future feature if that becomes desirable.